File: 1120-maximum-average-subtree/1120-maximum-average-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    //PAIR --> {SUM , totalNodes}
    double maxAvg = 0;
    pair<double,int> getAvg(TreeNode* node) {
        //base case
        if(!node) return {0,0};
        //left and right right recursive call
        pair leftAns = getAvg(node->left);
        pair rightAns = getAvg(node->right);
        //postOrderArea ---->>
        pair myAns = {0,0};        
    //total sum of my subtree
        myAns.first = leftAns.first + rightAns.first + node->val;
    //total nodes in my whole subtree
        myAns.second = leftAns.second + rightAns.second + 1;
    //maximuze Avg
        if(maxAvg< (double) myAns.first/myAns.second) maxAvg = (double) myAns.first/myAns.second;
        
        return myAns;
    }
    double maximumAverageSubtree(TreeNode* root) {
        getAvg(root);
        return maxAvg;
    }
};

File: 1120-maximum-average-subtree/1120-maximum-average-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    double maxAvg = 0;
    public class pair{
        double sum;
        int totalNodes;
        
        pair(double sum,int totalNodes) {
            this.sum = sum;
            this.totalNodes = totalNodes;
        }
    }
    public pair getMax(TreeNode node) {
        if(node==null) return new pair(0,0);
        
        pair leftAns = getMax(node.left);
        pair rightAns = getMax(node.right);
        
        pair myAns = new pair(0,0);
        myAns.sum = leftAns.sum + rightAns.sum + node.val;
        myAns.totalNodes = leftAns.totalNodes + rightAns.totalNodes + 1;
        if(myAns.sum/myAns.totalNodes > maxAvg) maxAvg = myAns.sum/myAns.totalNodes;
        
        return myAns;
        
    }
    public double maximumAverageSubtree(TreeNode root) {
        getMax(root);
        return maxAvg;
    }
}